pdf/pdfcanvas.cpp

```diff
@@ -773,46 +647,176 @@ QSGNode* PDFCanvas::updatePaintNode(QSGNode *node, QQuickItem::UpdatePaintNodeDa
         }
     }
 
+    return root;
+}
+
+void PDFCanvas::updatePolish()
+{
+    if (d->layoutScheduled) {
+        layout();
+    }
+
+    if (!d->flickable) {
+        return;
+    }
+
+    QQuickWindow * const window = this->window();
+
+    //Visible area equals flickable translated by contentX/Y
+    const QRectF visibleArea = mapRectFromItem(d->flickable, d->flickable->boundingRect());
+
+    //Loaded area equals visible area scaled to five times the size
+    const QRectF loadedArea = {
+        visibleArea.x() - visibleArea.width() * 2,
+        visibleArea.y() - visibleArea.height() * 2,
+        visibleArea.width() * 5,
+        visibleArea.height() * 5,
+    };
+    const QRect textureLimit = {
+        0, 0,
+        int(2.5 * qMin(window->width(), window->height())),
+        int(2.5 * qMin(window->width(), window->height()))
+    };
+    const float renderingRatio = float(d->renderWidth) / width();
+
+    struct PriorityRequest { int index; int width; QRect area; };
+
+    QVector<PriorityRequest> priorityRequests;
+    int currentPage = d->currentPage;
+    qreal maxVisibleArea = 0.;
+
+    for (int i = 0; i < d->pages.count(); ++i) {
+        PDFPage &page = d->pages[i];
+
+        page.pageIsInCacheArea = page.rect.intersects(loadedArea);
+
+        const bool pageIsInVisibleArea = page.rect.intersects(visibleArea);
+
+        // Current rendering in pixels is done with a width of
+        // d->renderWidth which can be different than actual width()
+        // when zooming.
+        QRect pageRect = {
+            0, 0, d->renderWidth, int(page.rect.height() * renderingRatio)
+        };
+
+        if (pageIsInVisibleArea) {
+            if (!page.linksLoaded)
+                d->document->requestLinksAtPage(i);
```

Comment:
This might not be a recent regression, but seems problematic call here.

For scalability testing, having pdf format 1.7 reference document, some 1300 pages. Opening that, quick scrolling to bottom and then manually upwards makes the application consume 100% for a long time. Added some debug to queue handling, looks like it's easily queueing a thousand or two jobs for links of the same pages. Request avoided only once a job has finished.

linksLoaded -> linksRequested or something like that?

Reply:
Done.